module/zfs/vdev_raidz.c

```diff
@@ -2969,6 +3037,14 @@ raidz_reflow_sync(void *arg, dmu_tx_t *tx)
 	 * real offset from the MOS?  Or rely on ditto blocks?
 	 */
 	vdev_t *vd = vdev_lookup_top(spa, vre->vre_vdev_id);
+
+	/*
+	 * Invalidate scratch object on first vre_offset_phys update.
+	 * Enable first metaslab.
```

Comment:
I think we can assert that progress has been made past the scratch size?

The code flow here of when we use the scratch object is a little hard for me to follow and verify that it's correct.  I wonder if we could be more explicit about invalidating the scratch object once we make progress past its size.

Reply:
I placed the logs to raidz_reflow_impl() and logged cases when device indexes and offsets are equal for reading and writing.
Had found that it does not work for now, because we can get sectors overwriting after the first vre_offset_phys update.
It is possible do not invalidate scrach object at all, at least until next expansion process, but it is needed to enable first metaslab.